src/python/pants/util/ordered_set.py

```diff
@@ -72,19 +74,28 @@ def __eq__(self, other: Any) -> bool:
             return NotImplemented
         return all(x == y for x, y in itertools.zip_longest(self._items, other._items))
 
-    def union(self, *others: Iterable[T]) -> "_AbstractOrderedSet[T]":
+    def __or__(self: _TAbstractOrderedSet, other: Iterable[T]) -> _TAbstractOrderedSet:  # type: ignore[override]
+        # Differences with AbstractSet: our set union forces "other" to have the same type. That is, while
+        # AbstractSet allows {1, 2, 3} | {(True, False)} resulting in Set[Union[int, Tuple[bool, bool]]], the analogous
+        # for descendants  of _TAbstractOrderedSet is not allowed.
+        # GOTCHA: given higher kinded type _TAbstractOrderedSet[S]:
```

Comment:
Ditto the below since [Mypy docs](https://mypy.readthedocs.io/en/stable/generics.html) and [PEP-484](https://www.python.org/dev/peps/pep-0484/) make no mention of HKTs. How about something like:
```
# GOTCHA: given _TAbstractOrderedSet[S] where:
# + T is a subclass of S: ...
# + S is a subclass of T: ...
```